src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -1569,6 +1569,352 @@ void C2_MacroAssembler::vinsert(BasicType typ, XMMRegister dst, XMMRegister src,
+    // from memory into vector using integral gather instruction.
+    for (int i = 0, j = 0; i < lane_count_subwords; i += lane_count_ints, j++) {
+      vpxor(xtmp2, xtmp2, xtmp2, int_vec_enc);
+      kxnorwl(gmask, gmask, gmask);
+      vgather_subword_avx3(elem_bt, xtmp2, base, offset_vec, idx_base, i,
+                           idx_vec, xtmp1, xtmp3, gmask, int_vec_enc);
+      if (vlen_enc == Assembler::AVX_512bit) {
+        // Case to handle 32 byte gather operation.
+        assert(int_vec_enc == Assembler::AVX_512bit, "");
+        // Appropriately permute 256 bit lane holding 16 shorts accumulated
+        // using 512 bit integral gather operation.
+        if (j > 0) {
+          vinserti64x4(dst, dst, xtmp2, j);
+        } else {
+          evmovdquq(dst, xtmp2, vlen_enc);
+        }
+      } else if (vlen_enc == Assembler::AVX_256bit) {
+        // Case to handle 16 byte gather operation.
+        if (int_vec_enc == Assembler::AVX_512bit) {
+          // All 16 short values are loaded in one short by 512 bit integral
+          // gather.
+          vmovdqu(dst, xtmp2);
+        } else {
+          assert(int_vec_enc == Assembler::AVX_256bit, "");
+          // Permute 8 short values loaded using 256 bit integral gather.
+          vinserti32x4(dst, dst, xtmp2, j);
+        }
+      } else if (vlen_enc == Assembler::AVX_128bit) {
+        if (j > 0) {
+          // We enter here only if maximum integer vector size is less than 256
+          // bits.
+          assert(int_vec_enc == Assembler::AVX_128bit, "");
+          vpslldq(xtmp2, xtmp2, 8, vlen_enc);
+        }
+        vpor(dst, dst, xtmp2, vlen_enc);
+      }
+    }
+  }
+
+  if (mask != knoreg) {
+    if (elem_bt == T_BYTE) {
+      evmovdqub(dst, mask, dst, false, vlen_enc);
+    } else {
+      assert(elem_bt == T_SHORT, "");
+      evmovdquw(dst, mask, dst, false, vlen_enc);
+    }
+  }
+}
+
+void C2_MacroAssembler::vgather8b(BasicType elem_bt, XMMRegister dst,
+                                  Register base, Register idx_base,
+                                  Register rtmp, int vlen_enc) {
+  vpxor(dst, dst, dst, vlen_enc);
+  if (elem_bt == T_SHORT) {
+    for (int i = 0; i < 4; i++) {
+      movl(rtmp, Address(idx_base, i * 4));
+      pinsrw(dst, Address(base, rtmp, Address::times_2), i);
+    }
+  } else {
+    assert(elem_bt == T_BYTE, "");
+    for (int i = 0; i < 8; i++) {
+      movl(rtmp, Address(idx_base, i * 4));
+      pinsrb(dst, Address(base, rtmp), i);
+    }
+  }
+}
+
+void C2_MacroAssembler::vgather8b_offset(BasicType elem_bt, XMMRegister dst,
+                                         Register base, Register idx_base,
+                                         Register offset, Register rtmp,
+                                         int vlen_enc) {
+  vpxor(dst, dst, dst, vlen_enc);
+  if (elem_bt == T_SHORT) {
+    for (int i = 0; i < 4; i++) {
+      movl(rtmp, Address(idx_base, i * 4));
+      addl(rtmp, offset);
```

Comment:
Can the `offset` not be added to `idx_base` before the loop?

Reply:
Or would that require too many registers?

Reply:
> Can the `offset` not be added to `idx_base` before the loop?

Offset needs to be added to each index element, please refer to API specification for details.
https://docs.oracle.com/en/java/javase/21/docs/api/jdk.incubator.vector/jdk/incubator/vector/ShortVector.html#fromArray(jdk.incubator.vector.VectorSpecies,short[],int,int[],int)

Reply:
Ah great, thanks for the link :blush: 
Can you put such equations in the code, using the register names?